### PR TITLE
gh-76785: Move the Cross-Interpreter Code to Its Own File

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This is partly to clear this stuff out of pystate.c, but also in preparation for moving some code out of _xxsubinterpretersmodule.c.

This change also moves this stuff to the internal API (new: Include/internal/pycore_crossinterp.h).  @vstinner did this previously and I undid it.  Now I'm re-doing it. :/

(This PR essentially only moves existing code between files.)

* Issue: gh-76785

### CODE REVIEW

File: Python/crossinterp.c [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+#include "pycore_pyerrors.h"      // _PyErr_Clear()
+#include "pycore_pystate.h"       // _PyInterpreterState_GET()
+#include "pycore_weakref.h"       // _PyWeakref_GET_REF()
+
+
+/**************************/
+/* cross-interpreter data */
+/**************************/
+
+/* cross-interpreter data */
```
Comment (Eclips4) [inline Python/crossinterp.c]: Looks like a duplicate of the above comment

Comment (ericsnowcurrently) [reply Python/crossinterp.c]: Thanks for pointing this out.  It was intentional, but the distinction wasn't particularly clear (obviously).  I've fix up the grouping, with explicit labels.

File: Include/internal/pycore_crossinterp.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    // cross-interpreter use, so it must be freed when
+    // _PyCrossInterpreterData_Release is called or the memory will
+    // leak.  In that case, at the very least this field should be set
+    // to PyMem_RawFree (the default if not explicitly set to NULL).
+    // The call will happen with the original interpreter activated.
+    xid_freefunc free;
+};
+
+PyAPI_FUNC(_PyCrossInterpreterData *) _PyCrossInterpreterData_New(void);
+PyAPI_FUNC(void) _PyCrossInterpreterData_Free(_PyCrossInterpreterData *data);
```
Comment (vstinner) [inline Include/internal/pycore_crossinterp.h:78]: If you don't need one of these functions in shared libraries (stdlib extensions), you can replace PyAPI_FUNC() with extern.

**Review by vstinner — APPROVED** [review]
LGTM.

File: Include/internal/pycore_crossinterp.h (hunk 2 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_crossinterp.h:78]: They're basically all used in _xxsubinterpretersmodule.c and _xxinterpchannelsmodule.c.

**Pull request merged**